### PR TITLE
Allow customizing Button type

### LINKED ISSUE #98: Component.Button stylesheet should inherit from InteractiveViewStyleSheet
The `Component.Button.Stylesheet` currently inherits from `BaseViewStyleSheet`, but the base view for the `Button` component is an `InteractiveView` so the stylesheet class should inherit from `InteractiveViewStyleSheet`.

A consequence of this bug is that you cannot set the `highlightColor` for a Button. That means that when you set custom `.highlighted` color for the button, it is overridden with the default `highlightColor` of the `InteractiveView`:

buttonbug

### PR DESCRIPTION
**Pull request by mluisbrown**
Also: Button should use InteractiveViewStyleSheet
Fixes #98 

Add a `type` to the Button component and allow this to be configured by the user. This allows using the `.custom` type, which is the only way to customize the button highlighted background colour without having a system fade applied. 

This was the actual cause of #98, not the incorrect style sheet type, but I fixed that also, just to be correct.

### CODE REVIEW

**Comment by mluisbrown** [comment]
> the UIButton should be dynamically recreated whenever the button type changes which is also a technique already used in several BentoKit & internal components.

Do you have an example?

**Comment by sergdort** [comment]
> using distinct reuse IDs
Yeah this should probably work

**Comment by sergdort** [comment]
@mluisbrown you just need to override `reuseId` property on the component. This is why it convenient for it to be instance property, @andersio 😜

**Comment by andersio** [comment]
To be clear, changing reuse identifier would solve the issue, provided that we also fix the adapters which currently does not handle component views with mismatching reuse identifiers as a result of silent updates (aka in-place reload). Just that I don't think it is a very compelling solution in _mitigating_ this class of issues, which is more about developer experience.

**Comment by mluisbrown** [comment]
@andersio I changed it so the Button is always recreated if the type changes.

File: BentoKit/BentoKit/Components/Button.swift [hunk 1]
```diff
@@ -61,25 +61,67 @@ extension Component.Button {
...
+                return _huggingConstraints
+            }
 
-        private lazy var strictConstraints: [NSLayoutConstraint] = [
-            button.leadingAnchor.constraint(equalTo: layoutMarginsGuide.leadingAnchor)
-                .withPriority(.defaultHigh),
-            layoutMarginsGuide.trailingAnchor.constraint(equalTo: button.trailingAnchor)
-                .withPriority(.defaultHigh)
-        ]
+            _huggingConstraints = [
```
Comment (andersio) [inline BentoKit/BentoKit/Components/Button.swift]: I guess we should probably create these as part of `setupLayout()` now.

File: BentoKit/BentoKit/Components/Button.swift [hunk 2]
```diff
@@ -61,25 +61,67 @@ extension Component.Button {
...
+                button = Button(type: buttonType)
+
+                _huggingConstraints = []
+                _strictConstraints = []
+
+                button.setContentHuggingPriority(.required, for: .vertical)
+                button.addTarget(self, action: #selector(buttonPressed), for: .touchUpInside)
+                setupLayout()
+
+                if hugsContent {
```
Comment (andersio) [inline BentoKit/BentoKit/Components/Button.swift]: Shouldn't we activate the strict constraints if we do not want to hug content?

File: BentoKit/BentoKit/Components/Button.swift [hunk 3]
```diff
@@ -61,25 +61,67 @@ extension Component.Button {
...
 
-        public let button = Button(type: .system).with {
-            $0.setContentHuggingPriority(.required, for: .vertical)
+        public var button = Button(type: .system)
+
+        fileprivate var buttonType: UIButton.ButtonType = .system {
+            didSet {
+                guard oldValue != buttonType else { return }
+
+                activityIndicator.removeFromSuperview()
```
Comment (andersio) [inline BentoKit/BentoKit/Components/Button.swift]: The logic here can be incorporated into `setupLayout()`, so we just need to invoke `setupLayout()` here to reconfigure as needed. Would centralize all these relevant bits too.

**Comment by mluisbrown** [comment]
@andersio comments addressed.

**Comment by mluisbrown** [comment]
Comments addressed @andersio

**Pull request merged**